### ISSUE TITLE
Application Insights loads different version of System.Data.SqlClient

### ISSUE THREAD

**Issue by swettstein** [Azure/azure-functions-host#3476]
I am having problems using a preview version of System.Data.SqlClient due to needing the AccessToken property so that my Azure Function can seamlessly connect to SqlAzure databases. I added a package reference to this version: 4.6.0-preview2-26905-02 based on [this conversation](https://github.com/dotnet/corefx/issues/13660). When trying to invoke the function, I get the following error:

AI (Internal): SqlClientDiagnosticSourceListener OnNext failed to call event handler. Error details 'System.InvalidCastException: [A]System.Data.SqlClient.SqlConnection cannot be cast to [B]System.Data.SqlClient.SqlConnection. Type A originates from 'System.Data.SqlClient, Version=4.5.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' in the context 'Default' at location 'D:\home\site\wwwroot\bin\System.Data.SqlClient.dll'. Type B originates from 'System.Data.SqlClient, Version=4.4.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' in the context 'Default' at location 'D:\Program Files (x86)\SiteExtensions\Functions\2.0.12050-alpha\64bit\System.Data.SqlClient.dll'.
   at Microsoft.ApplicationInsights.DependencyCollector.Implementation.SqlClientDiagnostics.SqlClientDiagnosticSourceListener.System.IObserver<System.Collections.Generic.KeyValuePair<System.String,System.Object>>.OnNext(KeyValuePair`2 evnt)'

It seems that the function runtime is loading ApplicationInsights into the Default context but it's outside of the deps.json so that can't even unify the different versions.

It will run locally in VS just fine because it doesn't load ApplicationInsights.

#### Investigative information

- Azure Function App version 2.0: 2.0.12095.0 (beta)
- Timestamp: 2018-09-17 15:40:11.184
- Invocation ID: operation_Id == 'GtvQCjJZX+I='
- Region: East US

#### Repro steps

Add PackageReference to the preview version of System.Data.SqlClient:

```xml
  <ItemGroup>
    <PackageReference Include="System.Data.SqlClient" Version="4.6.0-preview2-26905-02" />
  </ItemGroup>
```

#### Known workarounds

I can workaround it using a custom Assembly Resolver class and register it as an extension, but this is unnecessary code that should be handled by the host.

#### Related information 

Related issues [here](https://github.com/Azure/azure-functions-host/issues/2473) and [here](https://github.com/Azure/azure-webjobs-sdk/issues/1472). In my case, downgrading SqlClient to 4.4.0 will not work because I need to use the AccessToken property.

**Comment by jnevins-gcm** [comment]
@fabiocav

Does v2 actually look at the function app’s deps.json? We also noticed that unless we hard code in the platform of SqlClient to use, the runtime won’t even locate the platform specific assembly.

**Comment by jnevins-gcm** [comment]
What is the expected behavior in terms of the host leveraging the app’s deps.json file? How should we be handling platform specific dependencies?

Even if the restrictions by appinsights were lightened, we also encountered that
- our function app targets netstandard
- SqlClient has a shim assembly for netstandard (ends up in bin( and then platform specific implementations in the runtimes directory (ends up in bin/runtimes)

What expectations are there around how the host can/should find the correct platforms specific SqlClient (or any platform specific implementation assembly)?

**Comment by fabiocav** [comment]
To clarify, the restriction is on `System.Data.SqlClient`.

The target and platform specific dependencies are fine. The host will honor runtime/platform specific dependencies in the `bin/runtimes` folder, using the proper [RID](https://docs.microsoft.com/en-us/dotnet/core/rid-catalog) lookup and fallback logic (matching the .NET Core runtime behavior). So, for example, a dependency with versions in `runtimes/win-x86` and `runtimes/win-x64` would be loaded from the appropriate folder depending on the host environment/process.

The runtime uses the deps.json file (exclusively) in the CSX/dynamic compiled model. For pre-compiled, that assets are resolved ahead of time (at build or publish) the only the artifacts are used, with the probing, resolution and unification logic previously mentioned.

**Comment by jnevins-gcm** [comment]
Ok so that closes out the AppInsights question...but in troubleshooting SqlClient for AppInsights it seems that we ran into a broader, independent problem. 

At first we assumed the problem was just an issue with runtime specific packages in general 

We observed at least two problems -

1. When we have a transitive reference to a package with runtime specific assemblies...

AzureFunctionApp > CommonCode > RuntimeSpecificPackage 

Since AzureFunctionApp targets netstandard2.0 (the metadata extension generator breaks if you try to target netcoreapp2.0), by default the runtimes folder doesn’t end up in the AzureFunctionApp bin (I think maybe because PrivateAssets includes runtimes by default and a netstandard2.0 target framework build doesn’t think it needs runtime specific assemblies in the output directory even if CopyLocalLockFileAssemblies is true). 

2. If the RuntimeSpecificPackage has both a runtimes folder AND a netstandard2.0 shim dll which just throws NotSupportedExceptions, AND we add an explicit reference to RuntimeSpecificPackage from AzureFunctionApp so that the runtimes folder gets copied to the bin directory...

The Azure Functions host ends up trying to load the shim dll from the root of the bin directory and ignores the runtime specific dll in the runtimes directory
